ruma-appservice-api/src/lib.rs

```diff
@@ -4,6 +4,138 @@
 
 #![warn(missing_debug_implementations, missing_docs)]
 
+use serde::{Deserialize, Serialize};
+
 pub mod event;
 pub mod query;
 pub mod thirdparty;
+
+/// A namespace defined by an application service.
+///
+/// Used for [appservice registration](https://matrix.org/docs/spec/application_service/r0.1.2#registration).
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Namespace {
+    /// Whether this application service has exclusive access to events within this namespace.
+    pub exclusive: bool,
+
+    /// A regular expression defining which values this namespace includes.
+    pub regex: String,
+}
+
+impl Namespace {
+    /// Creates a new `Namespace` with the given exclusivity and regex pattern.
+    pub fn new<S: AsRef<str>>(exclusive: bool, regex: S) -> Self {
+        Namespace { exclusive, regex: regex.as_ref().into() }
+    }
+}
+
+/// Namespaces defined by an application service.
+///
+/// Used for [appservice registration](https://matrix.org/docs/spec/application_service/r0.1.2#registration).
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Namespaces {
+    /// Events which are sent from certain users.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub users: Vec<Namespace>,
+
+    /// Events which are sent in rooms with certain room aliases.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub aliases: Vec<Namespace>,
+
+    /// Events which are sent in rooms with certain room IDs.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub rooms: Vec<Namespace>,
+}
+
+impl Namespaces {
+    /// Creates a new `Namespaces` instance with the given `users`, `aliases`, and `rooms`
+    /// namespaces (none of them are explicitly required)
+    pub fn new(users: Vec<Namespace>, aliases: Vec<Namespace>, rooms: Vec<Namespace>) -> Self {
```

Comment:
I think in `Namespaces::new(vec![a], vec![b, c], vec![])` it's very unclear what each of the elements are (they're all the same type). I'd prefer if the constructor just set everything to empty `Vec`s (i.e. body can just be `Self::default()` after the above suggestion is applied) and then users can do

```rust
use assign::assign;

let namespaces = assign!(Namespaces::new(), {
    users: vec![a, b],
    aliases: vec![c],
});
```

Reply:
> it's very unclear what each of the elements are

Yeah, I would prefer a builder style for something like this:
```rust
Namespaces::new()
    .user(a)
    .user(b)
    .alias(c);
```

`assign` looks pretty good too.

Reply:
We use assign elsewhere too, I prefer it because it requires no additional code in our crates (which even if generated would increase compile times).